File: src/com/fictional_particle_sim/physicals/Barrier.java
package com.fictional_particle_sim.physicals;

import com.fictional_particle_sim.geometrics.BoundingBox;
import com.fictional_particle_sim.geometrics.Point;
import com.fictional_particle_sim.geometrics.Vector;

import java.awt.*;

public class Barrier {

    public Vector line;
    public Point center, topLeft, bottomRight;
    public double radius;
    public String shape;
    public Color color;

    public Barrier(Vector v, Color color) {
        line = v;
        this.color = color;
        shape = "LINE";

        center = new Point();
        topLeft = new Point();
        bottomRight = new Point();
    }
    public Barrier(Point center, double radius, Color color) {
        this.center = center;
        this.radius = radius;
        this.color = color;
        shape = "CIRCLE";

        line = new Vector();
        topLeft = new Point();
        bottomRight = new Point();

    }
    public Barrier(Point topLeft, Point bottomRight, Color color) {

        this.topLeft = new Point(Math.min(topLeft.x, bottomRight.x), Math.min(topLeft.y, bottomRight.y));
        this.bottomRight = new Point(Math.max(topLeft.x, bottomRight.x), Math.max(topLeft.y, bottomRight.y));
        this.color = color;
        this.shape = "RECTANGLE";

        line = new Vector();
        center = new Point();
    }

    public BoundingBox boundingBox() {

        return new BoundingBox(this);
    }
}

File: src/com/fictional_particle_sim/physicals/Barrier.kt
package com.fictional_particle_sim.physicals

import com.fictional_particle_sim.geometrics.BoundingBox
import com.fictional_particle_sim.geometrics.Point
import com.fictional_particle_sim.geometrics.Vector
import com.fictional_particle_sim.physicals.Shape.*
import java.awt.Color

class Barrier(var line: Vector = Vector(), var topLeft: Point = Point(), var bottomRight: Point = Point(), var center: Point = Point(), var radius: Double = 0.0, var shape: Shape, var color: Color = Color.CYAN) {


    fun boundingBox(): BoundingBox {
        val a: Point
        val b: Point
        when (shape) {
                LINE -> {
                    a = line.boundingBox().a
                    b = line.boundingBox().b
                }
                CIRCLE -> {
                    a = center - Point(radius, radius)
                    b = center + Point(radius, radius)
                }
                RECTANGLE -> {
                    a = topLeft
                    b = bottomRight
                }
            }
        return BoundingBox(a, b)
        }

    override fun toString(): String {
        return when (shape) {
            LINE -> "$shape, ${line.start}, ${line.end}"
            RECTANGLE -> "$shape, $topLeft, $bottomRight"
            CIRCLE -> "$shape, $center, $radius"
        }
    }
}


enum class Shape {
    LINE, CIRCLE, RECTANGLE
}